src/plugins/plugin-redis.js

```diff
@@ -0,0 +1,211 @@
+/**
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+var shimmer = require('shimmer');
+
+function createClientWrap(api, createClient) {
+  return function createClientTrace() {
+    var client = createClient.apply(this, arguments);
+    api.wrapEmitter(client);
+    return client;
+  };
+}
+
+// Used for redis version > 2.3
+function createStreamWrap(api, create_stream) {
+  return function create_stream_trace() {
+    if (!this.stream) {
+      Object.defineProperty(this, 'stream', {
+        get: function () { return this._google_trace_stream; },
+        set: function (val) {
+          api.wrapEmitter(val);
+          this._google_trace_stream = val;
+        }
+      });
+    }
+    return create_stream.apply(this, arguments);
+  };
+}
+
+// Used for redis version <= 2.3
+function streamListenersWrap(api, install_stream_listeners) {
+  return function install_stream_listeners_trace() {
+    api.wrapEmitter(this.stream);
+    return install_stream_listeners.apply(this, arguments);
+  };
+}
+
+function setupSpan(api, cmd, args, skipped_frames) {
+  var span = api.createChildSpan({
+    name: 'redis-' + cmd,
+    skipFrames: skipped_frames + 1
+  });
+  if (!span) {
+    return null;
+  }
+  span.addLabel('command', cmd);
+  if (api.enhancedDatabaseReportingEnabled()) {
+    span.addLabel('arguments', JSON.stringify(args));
+  }
+  return span;
+}
+
+function startSpanFromArguments(api, cmd, args, cb, send_command) {
+  if (!cmd || !args || typeof cmd !== 'string' || !Array.isArray(args) ||
+    (cb && typeof cb !== 'function')) {
+    return send_command(cmd, args, cb);
+  }
+  if (!cb) {
+    if (typeof args[args.length - 1] === 'function' ||
+        typeof args[args.length - 1] === 'undefined') {
+      cb = args.pop();
+    }
+  }
+  var span = setupSpan(api, cmd, args, 1);
+  if (!span) {
+    return send_command(cmd, args, cb);
+  }
+  return send_command(cmd, args, wrapCallback(api, span, cb));
+}
+
+function internalSendCommandWrap(api, internal_send_command) {
+  return function internal_send_command_trace(cmd, args, cb) {
+    if (arguments.length === 1 && typeof cmd === 'object') {
```

Comment:
In addition - is it possible to simplify this? It seems like the real reason for the `if` block is to correctly identify the command, arguments, and callback, if they are not provided directly as arguments. The remaining stuff in the `if` block can be done in `startSpanFromArguments` -- which also removed the need for a separate `setupSpan` object.